url7b90c1010b_DastanIqbal_HelloZMQwithJava_tgz-pJ8-ClientJ8/src/zmq/Poller.java

```diff
@@ -0,0 +1,277 @@
+    private boolean retired;
+
+    //  If true, thread is in the process of shutting down.
+    volatile private boolean stopping;
+    volatile private boolean stopped;
+    
+    private Thread worker;
+    private Selector selector;
+    final private String name;
+    
+    public Poller() {
+        this("poller");
+    }
+    
+    public Poller(String name_) {
+        
+        name = name_;
+        retired = false;
+        stopping = false;
+        stopped = false;
+        
+        fd_table = new HashMap<SelectableChannel, PollSet>();
+        try {
+            selector = Selector.open();
+        } catch (IOException e) {
+            throw new ZError.IOException(e);
+        }
+    }
+
+    public void destroy() {
+        
+        if (!stopped) {
+            try {
+                worker.join();
+            } catch (InterruptedException e) {
+            }
+            
+            try {
+                selector.close();
+            } catch (IOException e) {
+            }
+        }
+    }
+    public final void add_fd (SelectableChannel fd_, IPollEvents events_)
+    {
+        fd_table.put(fd_, new PollSet(events_));
+        
+        adjust_load (1);
+        
+    }
+    
+
+    public final void rm_fd(SelectableChannel handle) {
+        
+        fd_table.get(handle).cancelled = true;
+        retired = true;
+
+        //  Decrease the load metric of the thread.
+        adjust_load (-1);
+    }
+    
+
+    public final void set_pollin (SelectableChannel handle_)
+    {
+        register(handle_, SelectionKey.OP_READ, false);
+    }
+    
+
+    public final void reset_pollin (SelectableChannel handle_) {
+        register(handle_, SelectionKey.OP_READ, true);
+    }
+    
+    public final void set_pollout (SelectableChannel handle_)
+    {
+        register(handle_,  SelectionKey.OP_WRITE, false);
+    }
+    
+    public final void reset_pollout (SelectableChannel handle_) {
+        register(handle_, SelectionKey.OP_WRITE, true);
+    }
+
+    public final void set_pollconnect(SelectableChannel handle_) {
+        register(handle_, SelectionKey.OP_CONNECT, false);
+    }
+    
+    public final void set_pollaccept(SelectableChannel handle_) {
+        register(handle_, SelectionKey.OP_ACCEPT, false);        
+    }
+
+    private final void register (SelectableChannel handle_, int ops, boolean negate)
+    {
+        PollSet pollset = fd_table.get(handle_);
+        
+        if (negate) 
+            pollset.ops = pollset.ops &~ ops;
+        else
+            pollset.ops = pollset.ops | ops;
+        
+        if (pollset.key != null)
+            pollset.key.interestOps(pollset.ops);
+        else
+            retired = true;
+    }
+    
+    public void start() {
+        worker = new Thread(this, name);
+        worker.start();
+    }
+    
+    public void stop() {
+        stopping = true;
+        selector.wakeup();
+    }
+    
+// JDK6    @Override
+    public void run () {
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

The cyclomatic complexity of this method is 18. By comparison, 99% of the methods in the CodeGuru reference dataset have a lower cyclomatic complexity. This indicates the method has a high number of decisions and it can make the logic difficult to understand and test. We recommend that you simplify this method or break it into multiple methods.